redaxo/src/core/lib/util/markdown.php

```diff
@@ -38,6 +29,107 @@ public static function factory()
      */
     public function parse($code)
     {
-        return $this->parser->text($code);
+        $parser = new ParsedownExtra();
+        $parser->setBreaksEnabled(true);
+
+        return $parser->text($code);
+    }
+
+    /**
+     * Parses markdown code and extracts a table-of-content.
+     *
+     * @param string $code Markdown code
+     *
+     * @return array tupel of table-of-content and content
+     */
+    public function parseWithToc($code)
+    {
+        $parser = new rex_parsedown();
+        $parser->setBreaksEnabled(true);
+
+        $content = $parser->text($code);
+        $headers = $parser->headers;
+
+        $previous = 1;
+        $toc = '';
+
+        foreach ($headers as $header) {
+            $level = $header['level'];
+
+            if ($level > $previous) {
+                $toc .= "<ul>\n";
+                $previous = $level;
+            } elseif ($level < $previous) {
+                $toc .= "</li>\n";
+                $toc .= "</ul>\n";
+                $previous = $level;
+            } else {
+                $toc .= "</li>\n";
+            }
+
+            $toc .= "<li>\n";
+            $toc .= '<a href="#'.$header['id'].'">'.$header['text']."</a>\n";
+        }
+
+        for (; $previous > 1; --$previous) {
+            $toc .= "</li>\n";
+            $toc .= "</ul>\n";
+        }
+
+        return [$toc, $content];
+    }
+}
+
+/**
+ * @internal
+ */
+class rex_parsedown extends ParsedownExtra
+{
+    private $ids = [];
+
+    public $headers = [];
+
+    protected function blockHeader($line)
+    {
+        $block = parent::blockHeader($line);
+
+        return $this->handleHeader($block);
+    }
+
+    protected function blockSetextHeader($line, array $block = null)
+    {
+        $block = parent::blockSetextHeader($line, $block);
+
+        return $this->handleHeader($block);
+    }
+
+    private function handleHeader(array $block)
+    {
+        [$level] = sscanf($block['element']['name'], 'h%d');
+
+        if ($level < 2 || $level > 3) {
```

Comment:
Fände es gut, wenn man das Level selbst wählen kann.

`public function parseWithToc($code, $topLevel = 1, $depth = 6)`

Reply:
Die aktuelle Variante (h2 und h3) wäre dann `, 2, 1)` oder `, 2, 2)`?

Reply:
ˋ, 2, 3)ˋ vielleicht den Variablennamen anpassen

Reply:
Gerade wenn wir hier allgemeiner werden, stellt sich die Frage, wenn in der Markdown-Datei Level ausgelassen werden, was dann rauskommen sollte.

Beispiel:
```markdown
# Haupt

### Foo

### Bar

## Baz
```

Was würdest du hier erwarten? (Bei Ausgabe 1 bis 4)
Vielleicht das hier?

```html
<ul>
  <li><a>Haupt</a>
    <ul>
      <li>
        <ul>
          <li><a>Foo</a></li>
          <li><a>Bar</a></li>
        </ul>
      </li>
      <li><a>Baz</a></li>
    </ul>
  </li>
</ul>
```

Es würde also teils li-Elemente ohne Bezeichnung (ohne `<a>`) geben.
Oder die Struktur ganz ablehnen (Fehler), da nicht wohlgeformt?

Reply:
> Oder die Struktur ganz ablehnen (Fehler), da nicht wohlgeformt?

Fände ich gut. Wenn falsches Markdown notiert wird, wird es auch als nicht wohlgeformt angesehen und es bleibt wie es ist.

Reply:
> und es bleibt wie es ist.

Den Part verstehe ich nicht.

So richtig falsch ist es übrigens nicht unbedingt. Oder ist im Markdown-Standard festgelegt, dass auf eine h1 eine h2 zu folgen hat?
Es ist halt bloß für die Navi. ;)

Reply:
Sorry, wenn ich mich da in eure interne Diskussion einmische

h1 wäre für mich Titel des Addons
h2 die Kategorie / Hauptnavigation
h3 die Unternavigation

Reply:
Kurz … Ich fände es gut, wenn unser Parser die Struktur korrekt erwartet und es ansonsten ablehnt.

Reply:
@alexplusde So genau wollen wir das nicht definieren sondern eher allgemein halten. Was die h1, h2 etc genau ist, kann der Autor entscheiden. Wenn das mit den level noch kommt, wäre es recht flexibel.

Reply:
Warum sollte man da Komplexität einbauen, wenn's auch so geht? @tbaddade

Reply:
Ich spreche jetzt mal für das Doku-Team mit, aber eine gewisse Praxis hat sich ja bereits an einigen Stellen bewährt.

Reply:
> Warum sollte man da Komplexität einbauen, wenn's auch so geht

verstehe nicht was du meinst.

Reply:
Sich darauf zu einigen, dass diese und jene Überschriftentiefe eben die Navi erzeugen, und andere nicht, halte ich für ausreichend.

Reply:
Mit @gharlan hatte ich besprochen, dass in der package.yml angegeben .md-Files die h2 und h3 als Navi erstellt werden. [Das würde an dieser Stelle passieren - 10:46 Uhr](https://github.com/redaxo/redaxo/blob/3ac0caacb34cb8f64fff061310e659783e72e7de/redaxo/src/core/lib/be/controller.php#L433-L444)

Parsed man selber die Dateien, kann man aber selbst bestimmen aus welchen Headlines die TOC erstellt wird. So können Redakteurstexte auch im Frontend mit einer TOC erscheinen.